### PR TITLE
HPCC-24303 Improve capture of ESP stats

### PR DESCRIPTION
**Pull request by kenrowland**
Added support in ecm service definition to automatically generate execution profile histograms

Signed-off-by: Ken Rowland Kenneth.Rowland@lexisnexisrisk.com

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: tools/hidl/hidlcomp.cpp
Comment (afishbeck) [inline tools/hidl/hidlcomp.cpp:7138]: Interesting way of updating the higher level flag.  I wonder, is this faster or clearer than the more explicit:
```
if (!executionProfilingEnabled && serviceProfileExecutionEnabled)
    executionProfilingEnabled = true;
```
?

File: tools/hidl/hidlcomp.cpp
Comment (afishbeck) [inline tools/hidl/hidlcomp.cpp:7181]: Did you mean?:
```
   if (executionProfilingEnabled)
   {
        outs("#include \"espcommon.hpp\"\n");
        outs("#include \"jmetrics.hpp\"\n");
   }
```

File: esp/espcommon/espcommon.cpp
Comment (afishbeck) [inline esp/espcommon/espcommon.cpp:64]: Trivial but FYI we do have existing routines that do this, it would just mean using a StringArray for executionProfileDef instead of a std::vector.  And it would mean not needing to create the splitString function.

```
StringArray executionProfileDef ;
executionProfileDef.appendList(profileOptions.c_str(), ",");
```

**Review by afishbeck — COMMENTED** [review]
@kenrowland some fairly minor comments.

File: esp/espcommon/espcommon.cpp
Comment (kenrowland) [reply esp/espcommon/espcommon.cpp:34]: I don't see it being part of jmetrics at this time. If other components accept a string to define the histogram buckets, we can revisit moving it.

File: esp/espcommon/espcommon.cpp
Comment (kenrowland) [reply esp/espcommon/espcommon.cpp:64]: If it's OK, I'd like to stick with the std lib types.

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:397]: Changed type to size_t

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:7138]: Since this is build time, performance is not as critical, but I think the |= is faster. If you'd like the more explicit version, I can change it.

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:7181]: Yes, probably. Good catch, although not harmful. Better to fix for sure.

**Review by afishbeck — APPROVED** [review]

**Comment by kenrowland** [comment]
@richardkchapman @ghalliday Please merge

**Comment by afishbeck** [comment]
@kenrowland I would say that the title of your PR could be a lot more descriptive.  "Improve capture of ESP stats".  Is pretty generic and not actually what you are doing.  What you are really doing is adding support for the instrumentation of some metrics to ESDL / HIDL.

File: tools/hidl/hidlcomp.h
Comment (ghalliday) [inline tools/hidl/hidlcomp.h:1017]: typo: profiling

File: esp/espcommon/espcommon.cpp
Comment (ghalliday) [inline esp/espcommon/espcommon.cpp:48]: Clean up unused code.

**Review by ghalliday — COMMENTED** [review]
@kenrowland broadly lookgs good.  A few areas that need looking at
- it needs more information in the jira about exactly what has been implemented
- this should have a documentation impact.  A new executionProfilingOptions has been option to the service definitions, there needs to be a description of the syntax and the effect.
- Some concern that use of std::string is duplicating functions that already exist, or make function prototypes ineffient
- Some scope for reducing the size of the generated code.
- Cleaning up some out of date comments and warnings

**Comment by kenrowland** [comment]
@ghalliday Please review changes. I opened a sub-task Jira to which I'd like to move this. Once the review is done, I'd like to squash these changes and rebase them to another branch matching the new Jira (which is more descriptive). Then close this PR.

File: tools/hidl/hidlcomp.cpp
Comment (ghalliday) [inline tools/hidl/hidlcomp.cpp:5754]: Missed the comment above about combining the functions - possibly because it wasn't attached to a line that has changed.

File: tools/hidl/hidlcomp.cpp
Comment (ghalliday) [reply tools/hidl/hidlcomp.cpp:7128]: Missed this comment

**Review by ghalliday — COMMENTED** [review]
@kenrowland looking good.  Two missed comments from before (commoning up generated code and unused variables)

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:5754]: Expanded the comment above the loop to indicate that its initializing the histogram metric for each method with execution profiling enabled.

**Comment by kenrowland** [comment]
@ghalliday Once complete, to make the PR title more descriptive, I can either change the title when rebasing, or close this PR and make a new one under a different Jira (https://track.hpccsystems.com/browse/HPCC-28576). Or, we can just leave it alone and not worry about it.

**Comment by ghalliday** [comment]
@kenrowland I think you misunderstood my last comments.  There were 2 comments from previous PRs you misssed - it wasn't a comment about comments in the code.
You should change the commit message when squashing this PR - it should be something that makes sense on its own in a change log.  There is no need to open a new PR.

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:5752]: Converted to a function in espcommon. That is one of the intended goals, so this is a good first step. I will test the generated code to make sure it works in a functioning ESP.

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:7125]: Added -Wunused-variable to the compile options. Resulted in a number of other unused variable warnings that I also removed.

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp:7126]: Removed

File: tools/hidl/hidlcomp.cpp
Comment (kenrowland) [reply tools/hidl/hidlcomp.cpp]: I followed the standard used elsewhere in the code. If you want, I can build a string with the name of the method and throw that. You are right that it may leak, but it's probably not an issue since this is a tool that only runs at build time. Please let me know if you want that change before final approval.

**Review by kenrowland — COMMENTED** [review]
@ghalliday  I did misunderstand your last comments. The comments to which you referred were collapsed. I addressed them in the latest commit. Please see my latest comments. I'm not able to request a re-reviews, so please take a look.

File: esp/espcommon/espcommon.cpp
Comment (ghalliday) [inline esp/espcommon/espcommon.cpp:41]: minor: would avoid creating a std::string if this was a const char * parameter

File: tools/hidl/CMakeLists.txt
Comment (ghalliday) [inline tools/hidl/CMakeLists.txt:53]: The following are the preferred options:

```
  -Wall -Wextra -Wno-switch -Wno-unused-parameter -Werror -Wno-error=delete-non-virtual-dtor
```
so that warnings are reported as errors.  (e.g see CMakeList.txt for jlib)

File: tools/hidl/hidlcomp.cpp
Comment (ghalliday) [reply tools/hidl/hidlcomp.cpp:7111]: tmp is also unused and hasn't been reported.  (Because it has a non-trivial constructor.)

File: esp/espcommon/espcommon.cpp
Comment (ghalliday) [reply esp/espcommon/espcommon.cpp:41]: could also make it static if not exported.

**Review by ghalliday — COMMENTED** [review]
A few (hopefully) final comments

File: tools/hidl/CMakeLists.txt
Comment (kenrowland) [reply tools/hidl/CMakeLists.txt]: I added the options as indicated. There were a number of additional errors that had to be corrected. Please see hidllex.cpp. In that file a method and it's prototype were commented out since they weren't used. However, there are a number of instances in the code that use the definition that controlled its inclusion. Not knowing this part of the code, I decided to comment it out for now.

**Comment by kenrowland** [comment]
@AttilaVamos I am looking into this. It has to do with yylex generated code.

File: tools/hidl/hidlgram.y
Comment (ghalliday) [inline tools/hidl/hidlgram.y:1695]: trivial: Should have a break after this statement to stop coverity complaining.

**Review by ghalliday — APPROVED** [review]
Looks good.  One trivial comment.  Please sqaush.  Re-requesting tony to look at the changes since quite a lot has happened since he last reviewed.

File: esp/espcommon/espcommon.cpp
Comment (afishbeck) [inline esp/espcommon/espcommon.cpp:55]: A bit inconsistent with singular vs plural name for profilingOption(s).

File: tools/hidl/hidlcomp.cpp
Comment (afishbeck) [inline tools/hidl/hidlcomp.cpp:396]: For a string validation function, a comment showing the expected format would be helpful.

File: tools/hidl/hidlcomp.h
Comment (afishbeck) [inline tools/hidl/hidlcomp.h:1016]: You could save the qualified name in the EspMethodInfo class rather than building it every time.

**Review by afishbeck — COMMENTED** [review]
A few fairly trivial comments.

**Comment by kenrowland** [comment]
@afishbeck @ghalliday Comments addressed. There are three additional changes in the latest commit
- Cleaned up header file usage in espcommon
- Added statement to hidl lex file to prevent generation of yyunput to fix a build problem in CentOS (caused by new build flags)
- Added validation of execution profiling options
I have re-requested a review to make sure these are covered.

File: tools/hidl/hidlcomp.h
Comment (afishbeck) [inline tools/hidl/hidlcomp.h:1005]: You only call this once, but still prefer to code defensively.  Should check to see if executionProfilingHistogramVariableName is empty befroe appending, or if there were a reason to update then clear it.

File: tools/hidl/hidlcomp.h
Comment (afishbeck) [inline tools/hidl/hidlcomp.h:1017]: Curious why you can't simply?:

executionProfilingOptions = options;

**Review by afishbeck — COMMENTED** [review]
Some comments.

File: tools/hidl/hidlcomp.h
Comment (kenrowland) [reply tools/hidl/hidlcomp.h:1005]: Changed the method to only turn on profiling (once). There is no reason to support allowing it to be turned off.

File: tools/hidl/hidlcomp.h
Comment (kenrowland) [reply tools/hidl/hidlcomp.h:1017]: I generally am more explicit in areas like this so that when you are looking at the code you'd know that a std::string was being constructed. I changed it to be more concise.

File: tools/hidl/hidlcomp.h
Comment (kenrowland) [reply tools/hidl/hidlcomp.h:1020]: You are correct, I do remove the spaces before calling the validation. However, the function could be called by someone else. I removed the check and made a note that the caller must remove any spaces first.

File: tools/hidl/hidlcomp.cpp
Comment (afishbeck) [inline tools/hidl/hidlcomp.cpp:402]: This is correct, but this wasn't exactly the intention of my comment.  Caller must remove spaces in order for there to be consistency, not just to pass this test.

Your syntax accepts spaces, it's an internal detail that they are stripped out before use.  If another place calls this validate and it doesn't work because spaces aren't removed then it will be confusing for users.

**Review by afishbeck — APPROVED** [review]
Approved.  @kenrowland trivial comment explaining the intent of my prior comment.

**Comment by ghalliday** [comment]
@kenrowland this is failing the macos12 build for the following reasons:

```
/Users/runner/work/HPCC-Platform/HPCC-Platform/build/tools/hidl/hidllex.cpp:2117:16: error: function 'yyinput' is not needed and will not be emitted [-Werror,-Wunneeded-internal-declaration]
    static int yyinput (void)
               ^
/Users/runner/work/HPCC-Platform/HPCC-Platform/tools/hidl/hidlcomp.cpp:5537:15: error: format string is not a string literal [-Werror,-Wformat-nonliteral]
        voutf(fmt, args);
              ^~~
```

File: tools/hidl/hidlcomp.cpp
Comment (ghalliday) [inline tools/hidl/hidlcomp.cpp:4431]: Not sure how renaming these variables is related to fixing the warnings, and this seems to accidentally have a double rr in it (which makes it heard to read).

File: tools/hidl/hidlcomp.cpp
Comment (ghalliday) [inline tools/hidl/hidlcomp.cpp:429]: picky: easier to read if value is also declared here, rather than wondering why value might be used after the loop, and is it a problem it isn't initialised.

**Review by ghalliday — COMMENTED** [review]
A couple of minor comments.  Only one really needs fixing. (attrrib)

**Comment by ghalliday** [comment]
@kenrowland looks like you only addressed one of the warnings/errors from macos.  The other one is:

```
/Users/runner/work/HPCC-Platform/HPCC-Platform/tools/hidl/hidlcomp.cpp:5537:15: error: format string is not a string literal [-Werror,-Wformat-nonliteral]
        voutf(fmt, args);
              ^~~
```

It requires adding __attribute__((format(printf, ....)) to various functions.

**Comment by kenrowland** [comment]
@ghalliday @richardkchapman Please merge.

**Comment by ghalliday** [comment]
@kenrowland what documentation impact does this have?  No checkboxes were ticked, but there must be some.  @JamesDeFabia  where is the syntax for esdl files defined?

**Comment by kenrowland** [comment]
@ghalliday I don't think we are ready to add this to the documentation yet. It would be part of a comprehensive set of documentation on metrics. My plan was to have a separate Jira at a later date.

**Comment by ghalliday** [comment]
It is worth creating a jira for the documentation, and including the information from each stage as it is implemented, to make sure that information isn't lost.

**Pull request merged**
